tools/pythonpkg/src/pyconnection.cpp

```diff
@@ -325,10 +249,18 @@ unique_ptr<DuckDBPyRelation> DuckDBPyConnection::FromArrowTable(const py::object
 		throw std::runtime_error("Only arrow tables supported");
 	}
 
-	auto my_arrow_table = new PythonTableArrowArrayStream(table);
-	string name = "arrow_table_" + PtrToString((void *)my_arrow_table);
-	return make_unique<DuckDBPyRelation>(
-	    connection->TableFunction("arrow_scan", {Value::POINTER((uintptr_t)my_arrow_table)})->Alias(name));
+	unique_ptr<PythonTableArrowArrayStreamFactory> stream_factory =
+	    make_unique<PythonTableArrowArrayStreamFactory>(table);
+	string name = "arrow_table_" + PtrToString((void *)&table);
+	ArrowArrayStream *(*stream_factory_produce)(uintptr_t factory) = PythonTableArrowArrayStreamFactory::Produce;
+	auto rel = make_unique<DuckDBPyRelation>(
+	    connection
+	        ->TableFunction("arrow_scan", {Value::POINTER((uintptr_t)stream_factory.get()),
+	                                       Value::POINTER((uintptr_t)stream_factory_produce)})
+	        ->Alias(name));
+	registered_arrow_factory[PtrToString((void *)&table)] = move(stream_factory);
```

Comment:
Are we keeping a proper reference to the input arrow table here? Could you add a test similar to `test_pandas_unregister2` (from [here](https://github.com/duckdb/duckdb/blob/master/tools/pythonpkg/tests/pandas/test_pandas_unregister.py)) where we try to use the arrow table after explicitly deleting the reference we have to it in Python?

Reply:
Yes, the factory holds a ref to it, that's an excellent idea!